### ISSUE TITLE
Excessive geometry calls

### ISSUE THREAD

**Issue by stndart** [stndart/2d-velocity-pathfinding#14]
How to reproduce:
launch main.py from b42a9870b2f7a940549e6fb3b36887ac4c241da3

**Comment by stndart** [comment]
Profiler output:
```

Building graph took  9.72s
         17329309 function calls (17309979 primitive calls) in 9.720 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.010    0.010    9.720    9.720 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:74(build_graph_on_quadtree)
     1085    0.005    0.000    8.637    0.008 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:431(__call__)
     1085    0.007    0.000    8.619    0.008 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:101(add_edge)
     2219    0.135    0.000    8.378    0.004 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:78(add_vertex)
    71331    0.092    0.000    8.365    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:5(get_matching_item)
   142629    0.835    0.000    8.228    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:6(<genexpr>)
    72465    0.046    0.000    8.198    0.000 {built-in method builtins.next}
  1382450    1.187    0.000    7.413    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:37(__eq__)
  1382450    2.081    0.000    6.137    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:166(__eq__)
  3010179    0.823    0.000    1.989    0.000 {built-in method builtins.abs}
  1567641    1.401    0.000    1.782    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:157(__sub__)
  1475473    0.978    0.000    1.165    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:136(__abs__)
     1417    0.004    0.000    1.034    0.001 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:49(check_collisions)
6229/1417    0.022    0.000    0.859    0.001 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:226(get_collision_candidates)
15653/6259    0.035    0.000    0.832    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:507(has_intersect)
30572/26018    0.042    0.000    0.733    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:240(has_intersect)
  1416286    0.557    0.000    0.658    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:9(is_close)
    29054    0.177    0.000    0.652    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:249(_intersects_line)
     6125    0.019    0.000    0.624    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:515(<listcomp>)
  1652877    0.402    0.000    0.402    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:84(__init__)
    58108    0.111    0.000    0.318    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:222(distance)
  2961916    0.191    0.000    0.191    0.000 {built-in method builtins.isinstance}
  1475473    0.187    0.000    0.187    0.000 {built-in method math.sqrt}
    33830    0.052    0.000    0.153    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:175(__init__)
     6290    0.025    0.000    0.152    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:478(edges)
     1524    0.004    0.000    0.125    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:398(has_intersect)
     1518    0.003    0.000    0.095    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:410(<listcomp>)
    47174    0.048    0.000    0.062    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:139(__mul__)
    69112    0.040    0.000    0.052    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:56(__hash__)
     1423    0.004    0.000    0.042    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\sprites\sprite.py:68(make_sprite)
     2219    0.016    0.000    0.034    0.000 {method 'update' of 'dict' objects}
     1518    0.004    0.000    0.026    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:374(edges)
     4072    0.007    0.000    0.024    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:516(<listcomp>)
     2744    0.005    0.000    0.023    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:181(copy)
    17719    0.016    0.000    0.021    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:154(__add__)
    17770    0.015    0.000    0.020    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:151(__truediv__)
     9424    0.009    0.000    0.019    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:486(contains)
    34801    0.017    0.000    0.019    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:59(__eq__)
    56150    0.016    0.000    0.016    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:12(sign)
     1423    0.004    0.000    0.015    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\sprites\sprite.py:6(__init__)
     6295    0.011    0.000    0.015    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:468(corners)
    35641    0.014    0.000    0.014    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:43(__init__)
     1085    0.002    0.000    0.012    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:304(dispatch)
   138224    0.012    0.000    0.012    0.000 {built-in method builtins.id}
     2219    0.002    0.000    0.011    0.000 {method 'add' of 'set' objects}
     1423    0.002    0.000    0.010    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:70(mean_points)
   355/49    0.002    0.000    0.009    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:121(find_adjacent)
     2170    0.007    0.000    0.009    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:334(dispatch_iter)
     2170    0.004    0.000    0.008    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:24(add_edge_to)
     5590    0.004    0.000    0.006    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:88(copy)
        1    0.000    0.000    0.005    0.005 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:60(merge_vertexes)
     1423    0.001    0.000    0.005    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:68(sum_points)
       49    0.000    0.000    0.005    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:57(get_matching_item)
       98    0.000    0.000    0.005    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:58(<genexpr>)
    10353    0.005    0.000    0.005    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:61(iter_children)
     1423    0.001    0.000    0.004    0.000 {built-in method _functools.reduce}
     1085    0.001    0.000    0.004    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:42(distance)
     1372    0.001    0.000    0.003    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:69(<lambda>)
      714    0.001    0.000    0.002    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:72(which_corner)
     2268    0.002    0.000    0.002    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:34(__hash__)
    13409    0.002    0.000    0.002    0.000 {built-in method builtins.any}
     5444    0.002    0.000    0.002    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:193(vertexes)
      714    0.001    0.000    0.001    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:58(childless)
    10283    0.001    0.000    0.001    0.000 {method 'update' of 'set' objects}
     1085    0.001    0.000    0.001    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:432(<listcomp>)
      142    0.000    0.000    0.001    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:29(opposite_corner)
     1087    0.001    0.000    0.001    0.000 C:\Python311\Lib\site-packages\multipledispatch\variadic.py:32(isvariadic)
     5791    0.001    0.000    0.001    0.000 {built-in method builtins.len}
      426    0.000    0.000    0.001    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:30(<genexpr>)
     1085    0.000    0.000    0.001    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:252(ordering)
     1085    0.001    0.000    0.001    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:426(__get__)
     1089    0.000    0.000    0.000    0.000 {built-in method builtins.all}
      714    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:59(<listcomp>)
  365/153    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:169(recursive_parents)
        5    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:19(build_vertexes_from_rect)
      284    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:21(opposite_direction)
     30/6    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:241(get_quad_tree)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\dispatcher.py:259(reorder)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:109(ordering)
       55    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:30(__init__)
        5    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:24(<listcomp>)
       30    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:55(contains)
        6    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:106(has_intersect)
      444    0.000    0.000    0.000    0.000 {method 'index' of 'list' objects}
        6    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:378(contains)
    40/12    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\quadtree.py:67(dfs)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:115(<listcomp>)
        4    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:97(edge)
       24    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:363(area)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:33(_toposort)
       55    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:9(__init__)
        5    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:459(vertexes)
        6    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:9(supercedes)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.print}
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:76(ambiguities)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:79(<genexpr>)
       51    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:91(vertexes)
       95    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:71(ambiguous)
        3    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:120(<genexpr>)
        1    0.000    0.000    0.000    0.000 {built-in method fromkeys}
        5    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\buildgraph.py:26(<listcomp>)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:39(consistent)
        3    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:63(<genexpr>)
        2    0.000    0.000    0.000    0.000 {built-in method time.time}
       12    0.000    0.000    0.000    0.000 {built-in method builtins.hash}
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:68(reverse_dict)
        3    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:52(<genexpr>)
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:90(groupby)
        2    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\conflict.py:120(<listcomp>)
        2    0.000    0.000    0.000    0.000 {method 'popitem' of 'collections.OrderedDict' objects}
        2    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\geometry.py:357(vertexes)
        4    0.000    0.000    0.000    0.000 {method 'get' of 'dict' objects}
        2    0.000    0.000    0.000    0.000 {method 'items' of 'collections.OrderedDict' objects}
        1    0.000    0.000    0.000    0.000 c:\Users\sstnd\documents\Projects\2d-speed-nav\backend\pathfinding\graph.py:75(__init__)
        3    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 C:\Python311\Lib\site-packages\multipledispatch\utils.py:51(<genexpr>)


approx is 30x12=360
Searching path took  1.49s

```

**Comment by stndart** [comment]
So the main issue was that vertexes were added multiple times, and each time all edges were recalculated (even if they were already present). Adding a check (if vertex is already present) reduced graph building times tenfold.
Closed in 5e3323ee34cb42906eff7a5d84e72d20d8d2191c
